spark/src/main/scala/geotrellis/spark/op/focal/FocalOperation.scala

```diff
@@ -10,160 +12,117 @@ import org.apache.spark.SparkContext._
+        case BottomRight => 
+          GridBounds(0, 0, m-1, m-1)
+        case BottomLeft => 
+          GridBounds(m+cols, 0, m+cols+m-1, m+rows+m-1)
+      }
     }
-
-    (y, x)
   }
 
-}
-
-trait FocalOperation[K] extends RasterRDDMethods[K] {
-
-  val _sc: SpatialComponent[K]
-
-  def zipWithNeighbors: RDD[(K, Tile, TileNeighbors)] = {
-    val sc = rasterRDD.sparkContext
-    val bcMetadata = sc.broadcast(rasterRDD.metaData)
-
-    val tilesWithNeighborIndices = rasterRDD.map { case (key, tile) =>
-      val metadata = bcMetadata.value
-      val gridBounds = metadata.gridBounds
-      val SpatialKey(col, row) = key
-      (key, tile, FocalOperation.getNeighborCoordinates(gridBounds, col, row))
-    }
+  import Direction._
+  
+  def apply[K: SpatialComponent: ClassTag](rdd: RDD[(K, Tile)], neighborhood: Neighborhood, opBounds: Option[GridBounds] = None)
+      (calc: (Tile, Neighborhood, Option[GridBounds]) => Tile): RDD[(K, Tile)] = {
+
+    val bounds = opBounds.getOrElse(GridBounds(Int.MinValue, Int.MinValue, Int.MaxValue, Int.MaxValue))
+    val m: Int = neighborhood.extent // how many pixels we need for the margin        
+    
+    rdd
+      .flatMap { case record @ (key, tile) =>
+        val SpatialKey(col, row) = key.spatialComponent
+        val slivers = new Array[(K, (Direction, Tile))](9)
+        var sliverCount = 0
+        // Tile.crop is inclusive on the max bounds, adjusting offsets to account for that
+        val cols = tile.cols-1
+        val rows = tile.rows-1
+        val mm = m - 1
+        
+        // ex: adding "TopLeft" corner of this tile to contribute to "TopLeft" tile at key
+        def addSlice(spatialKey: SpatialKey, direction: => Direction, sliver: => Tile) {
+          if (bounds.contains(spatialKey.col, spatialKey.row)) {
+            slivers(sliverCount) = key.updateSpatialComponent(spatialKey) -> (direction, sliver.toArrayTile) // force tile crop
+            sliverCount += 1
+          }
+        }
 
-    getNeighbors(tilesWithNeighborIndices)
+        // ex: A tile that contributes to the top (tile above it) will give up it's top slice, which will be placed at the bottom of the target focal window        
+        addSlice(SpatialKey(col,row), Center, tile)
+      
+        addSlice(SpatialKey(col-1, row), Left, tile.crop(0, 0, mm, rows))
+        addSlice(SpatialKey(col+1, row), Right, tile.crop(cols-mm, 0, cols, rows))
+        addSlice(SpatialKey(col, row-1), Top, tile.crop(0, 0,  cols, mm))
+        addSlice(SpatialKey(col, row+1), Bottom, tile.crop(0, rows-mm, cols, rows))
+
+        addSlice(SpatialKey(col-1, row-1), TopLeft, tile.crop(0, 0, mm, mm))
+        addSlice(SpatialKey(col+1, row-1), TopRight, tile.crop(cols-mm, 0, cols, mm))
+        addSlice(SpatialKey(col+1, row+1), BottomRight, tile.crop(cols-mm, rows-mm, cols, rows))
+        addSlice(SpatialKey(col-1, row+1), BottomLeft, tile.crop(0, rows-mm, mm, rows))
+
+        slivers.take(sliverCount)
```

Comment:
Use `ListBuffer` here instead, so you don't have to keep track of `sliverCount` and do this `take`?